---
title: "BGC exploration"
author: "Daniel Martinez"
format:
  html:
    df-print: paged
    code-fold: true
    code-tools: true
  pdf: 
    code-fold: true
    code-tools: true
editor: visual
---

# Background

This would be the part where the **background** of the project goes!

# Data exploration

## Reading the data

Here we are going to read the data we have

```{r}
#| warning: false
#| label: load-packages


library(tidyverse)
library(readr)
library(readxl)
library(cowplot)
```

```{r}
#| warning: false
#| label: read-data

# read the main dataset of BGCs
bgc = read_csv("../data/BGCs/antismash_summary.csv")

metadata = read_excel("../data/MAIN_metadata.xlsx", 
                       sheet = "metadata")
```

Now that we have the data generated, we can take a look at it:

```{r}
head(bgc)
```

From this table, the column type refers to...

## phylogroup exploration

This is the part where we start exploring the phylogroups and its relevance for the BGCs.

```{r}
phylo = 
  metadata %>% 
  filter(Origin %in% c('AUS', 'ECOREF')) %>% 
  filter(Discard == 'No') %>% 
  select(fasta, Broadphenotype, phylogroup) %>% 
  distinct(fasta, .keep_all = TRUE) %>% 
  mutate(genome = str_sub(fasta, 
                          start = 1, end = -7)) %>% 
  select(-fasta)
  

bgc_extended = bgc %>% 
  left_join(phylo) 
```

```{r}
#| label: fig-comp-paths-histo
#| fig-cap: Gene length for phylogroup

bgc_extended %>% 
  filter(start != 0 ) %>%
  # filter(type == 'NRPS') %>% 
  mutate(gene_length = end - start) %>% 
  mutate(ID = paste(genome, type, sep = '_')) %>% 
  mutate(type = as.factor(type)) %>% 
  ggplot(aes(y = phylogroup, 
             x = gene_length, 
             fill = phylogroup)) +
  geom_boxplot(show.legend = T) +
  geom_point(alpha = 0.5,
             show.legend = F) +
  theme_cowplot(15)
```

As we see in @fig-comp-paths-histo, we see that different phylogropus have different gene lenghts in the cohort.
